mlflow/server/otel_api.py

```diff
@@ -95,33 +98,70 @@ async def export_traces(
             detail="Invalid OpenTelemetry protobuf format",
         )
 
-    mlflow_spans = []
+    # Group spans by trace_id to support BatchSpanProcessor
+    # log_spans requires all spans in a batch to have the same trace_id
+    spans_by_trace_id = defaultdict(list)
     for resource_span in parsed_request.resource_spans:
         for scope_span in resource_span.scope_spans:
             for otel_proto_span in scope_span.spans:
                 try:
                     mlflow_span = Span.from_otel_proto(otel_proto_span)
-                    mlflow_spans.append(mlflow_span)
+                    spans_by_trace_id[mlflow_span.trace_id].append(mlflow_span)
                 except Exception:
                     raise HTTPException(
                         status_code=422,
                         detail="Cannot convert OpenTelemetry span to MLflow span",
                     )
 
-    if mlflow_spans:
+    if spans_by_trace_id:
         store = _get_tracking_store()
 
+        # Helper function to log spans for a single trace
+        def log_trace_spans(trace_id, trace_spans) -> str:
+            store.log_spans(x_mlflow_experiment_id, trace_spans)
+            return trace_id
+
+        errors = {}
         try:
-            store.log_spans(x_mlflow_experiment_id, mlflow_spans)
-        except NotImplementedError:
+            # Log each trace's spans in parallel using ThreadPoolExecutor
+            with ThreadPoolExecutor(
+                max_workers=MLFLOW_OTEL_ENDPOINT_TRACE_LOGGING_MAX_WORKERS.get(),
+                thread_name_prefix="MlflowOTelEndpointTraceLoggingWorker",
+            ) as executor:
+                # Submit all trace logging tasks
+                future_to_trace = {
+                    executor.submit(log_trace_spans, trace_id, trace_spans): trace_id
+                    for trace_id, trace_spans in spans_by_trace_id.items()
+                }
+
+                # Wait for all tasks to complete and check for errors
+                for future in as_completed(future_to_trace):
+                    trace_id = future_to_trace[future]
+                    try:
+                        future.result()
+                    except NotImplementedError:
+                        store_name = store.__class__.__name__
+                        raise HTTPException(
```

Comment:
when an error is thrown, does `ThreadPoolExecutor` cancel pending jobs? or wait for them to finish, then throw the error?